librz/core/cmeta.c

```diff
@@ -363,3 +363,128 @@ RZ_IPI void rz_core_meta_editor(RzCore *core, RzAnalysisMetaType mtype, ut64 add
 		free(out);
 	}
 }
+
+/**
+ * RzCoreMetaString
+ * @{
+ */
+
+static bool meta_string_8bit_add(RzCore *core, ut64 addr, size_t limit, ut8 **name, size_t *name_len) {
+	rz_return_val_if_fail(limit && name && name_len, false);
+	*name = malloc(limit + 1);
+	if (!*name) {
+		return false;
+	}
+	if (!rz_io_read_at(core->io, addr, *name, limit)) {
+		free(*name);
+		return false;
+	}
+	(*name)[limit] = '\0';
+	*name_len = strlen((char *)*name);
+	return true;
+}
+
+static bool meta_string_guess_add(RzCore *core, ut64 addr, size_t limit, ut8 **name, size_t *name_len, RzDetectedString **ds, RzStrEnc encoding) {
+	rz_return_val_if_fail(limit && name && name_len && ds, false);
+	*name = malloc(limit + 1);
+	if (!*name) {
+		return false;
+	}
+	RzBinFile *bf = rz_bin_cur(core->bin);
+	RzBinObject *obj = rz_bin_cur_object(core->bin);
+	if (!bf || !obj) {
+		free(*name);
+		return false;
+	}
+	bool big_endian = obj ? rz_bin_object_is_big_endian(obj) : RZ_SYS_ENDIAN;
+	RzUtilStrScanOptions scan_opt = {
+		.buf_size = 2048,
+		.max_uni_blocks = 4,
+		.min_str_length = 4,
+		.prefer_big_endian = big_endian
+	};
+	RzList *str_list = rz_list_new();
+	if (!str_list) {
+		free(*name);
+		return false;
+	}
+	ut64 paddr = rz_io_v2p(core->io, addr);
+	int count = rz_scan_strings(bf->buf, str_list, &scan_opt, paddr, paddr + limit, encoding);
+	if (count <= 0) {
+		rz_list_free(str_list);
+		free(*name);
+		return false;
+	}
+	*ds = rz_list_first(str_list);
+	rz_list_free(str_list);
+	rz_str_ncpy(*((char **)name), (*ds)->string, limit);
+	(*name)[limit] = '\0';
+	return true;
+}
+
+/**
+ * \brief add a string to RzCore
+ *
+ * \param core RzCore of core that will be add to
+ * \param addr string's address
+ * \param size string's max size
+ * \param encoding string's encoding
+ * \param name string's value, or null that will be autodetect at \p addr
+ * \return is add successful?
+ */
+RZ_API bool rz_core_meta_string_add(RzCore *core, ut64 addr, ut64 size, RzStrEnc encoding, RZ_NULLABLE const char *name) {
+	char *guessname = NULL;
+	size_t name_len = 0;
+	ut64 limit = size ? size : core->blocksize;
+	size_t n = 0;
+	bool result = false;
+	if (encoding == RZ_STRING_ENC_8BIT || encoding == RZ_STRING_ENC_UTF8) {
+		if (!meta_string_8bit_add(core, addr, limit, (ut8 **)&guessname, &name_len)) {
+			goto out;
+		}
+		n = size == 0 ? name_len + 1 : size;
+	} else {
+		RzDetectedString *ds = NULL;
+		if (!meta_string_guess_add(core, addr, limit, (ut8 **)&guessname, &name_len, &ds, encoding)) {
+			goto out;
+		}
+		if (!ds) {
+			goto out;
+		}
+		encoding = ds->type;
+		n = ds->size;
+	}
+	if (!name) {
+		result = rz_meta_set_with_subtype(core->analysis, RZ_META_TYPE_STRING, encoding, addr, n, guessname);
+	} else {
+		result = rz_meta_set_with_subtype(core->analysis, RZ_META_TYPE_STRING, encoding, addr, n, name);
+	}
+out:
+	free(guessname);
+	return result;
+}
+
+/**
+ * \brief add a pascal string to RzCore
+ * \see rz_core_meta_string_add
```

Comment:
please just copy the things from the `rz_core_meta_string_add` doxygen doc, otherwise if/when `rz_core_meta_string_add` is removed or the parameters changes we will for sure forget to update these as well.